src/nupic/algorithms/SDRClassifier.cpp

```diff
@@ -0,0 +1,741 @@
+        Real64 actValue, bool category, bool learn, bool infer,
+        ClassifierResult* result)
+      {
+        // save the offset between recordNum and learnIteration_ if this
+        // was not set (first call to compute)
+        if (!recordNumMinusLearnIterationSet_)
+        {
+          recordNumMinusLearnIteration_ = recordNum - learnIteration_;
+          recordNumMinusLearnIterationSet_ = true;
+        }
+
+        // update learnIteration_
+        learnIteration_ = recordNum - recordNumMinusLearnIteration_;
+
+        // update pattern history
+        patternNZHistory_.emplace_back(patternNZ.begin(), patternNZ.end());
+        iterationNumHistory_.push_back(learnIteration_);
+        if (patternNZHistory_.size() > maxSteps_)
+        {
+          patternNZHistory_.pop_front();
+          iterationNumHistory_.pop_front();
+        }
+
+        // if input pattern has greater index than previously seen, update 
+        // maxInputIdx and augment weight matrix with zero padding
+        if (patternNZ.size() > 0)
+        {
+          UInt maxInputIdx = *max_element(patternNZ.begin(), patternNZ.end());
+          if (maxInputIdx > maxInputIdx_)
+          {
+            maxInputIdx_ = maxInputIdx;
+            for (const auto& step : steps_)
+            {
+              Matrix& weights = weightMatrix_.at(step);
+              weights.resize(maxInputIdx_ + 1, maxBucketIdx_ + 1);
+            }
+          }
+        }
+
+        // if in inference mode, compute likelihood and update return value
+        if (infer)
+        {
+            infer_(patternNZ, bucketIdx, actValue, result);
+        }
+
+        // update weights if in learning mode
+        if (learn)
+        {
+          // if bucket is greater, update maxBucketIdx_ and augment weight
+          // matrix with zero-padding
+          if (bucketIdx > maxBucketIdx_) 
+          {
+            maxBucketIdx_ = bucketIdx;
+            for (const auto& step : steps_)
+            {
+              Matrix& weights = weightMatrix_.at(step);
+              weights.resize(maxInputIdx_ + 1, maxBucketIdx_ + 1);
+            }
+          }
+
+          // update rolling averages of bucket values
+          while (actualValues_.size() <= maxBucketIdx_)
+          {
+            actualValues_.push_back(0.0);
+            actualValuesSet_.push_back(false);
+          }
+          if (!actualValuesSet_[bucketIdx] || category)
+          {
+            actualValues_[bucketIdx] = actValue;
+            actualValuesSet_[bucketIdx] = true;
+          } else {
+            actualValues_[bucketIdx] =
+                ((1.0 - actValueAlpha_) * actualValues_[bucketIdx]) +
+                (actValueAlpha_ * actValue);
+          }
+
+          // compute errors and update weights
+          deque<vector<UInt>>::const_iterator patternIteration =
+              patternNZHistory_.begin();
+          for (deque<UInt>::const_iterator learnIteration =
```

Comment:
Maybe use `auto` for these iterators? Your call, really, but I think iterators are one of the best places for `auto`.